frame/executive/src/lib.rs

```diff
@@ -360,12 +362,32 @@ where
 		sp_tracing::enter_span!( sp_tracing::Level::TRACE, "finalize_block" );
 		<frame_system::Module<System>>::note_finished_extrinsics();
 		let block_number = <frame_system::Module<System>>::block_number();
-		<frame_system::Module<System> as OnFinalize<System::BlockNumber>>::on_finalize(block_number);
-		<AllModules as OnFinalize<System::BlockNumber>>::on_finalize(block_number);
+
+		Self::idle_and_finalize_hook(block_number);
 
 		<frame_system::Module<System>>::finalize()
 	}
 
+	fn idle_and_finalize_hook(block_number: NumberFor<Block>) {
+		let weight =  <frame_system::Module<System>>::block_weight();
+		let max_weight =  <
+			System::BlockWeights as frame_support::traits::Get<_>
+			>::get().max_block;
+		let mut remaining_weight = max_weight.saturating_sub(weight.total());
+
+		if remaining_weight > 0 {
+			let mut used_weight = <frame_system::Module<System> as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight);
+			remaining_weight = remaining_weight.saturating_sub(used_weight);
+			used_weight = <
+				AllModules as OnIdle<System::BlockNumber>
+				>::on_idle(block_number, remaining_weight).saturating_add(used_weight);
```

Comment:
how is remaining weight getting updated here for each of the hooks? this looks wrong

Reply:
i found the magic in the tuple impl